File: 异步子流程模板.js
/**
 * 小瓶RPA异步子流程模板
 * 
 * 功能说明：此脚本演示了RPA中如何使用异步子流程
 * 通过这个示例，您可以学习如何在主流程中调用同步和异步子流程，并处理子流程的返回结果
 */

// 引入小瓶RPA的核心库，获得对RPA功能的访问权限
const pbottleRPA = require('./pbottleRPA.js')

// 集中注册全局变量，其他地方使用不用加 global 前缀
global.global_processName = '小瓶RPA——XXXX流程模板'   // 定义全局流程名称变量
global.global_startTime = pbottleRPA.utils.getTime()   // 获取并定义全局开始时间变量

// 主流程函数，使用async关键字支持异步操作，使所有子流程序列化
async function main() {

    // 在日志中记录主流程开始信息，包括开始时间和流程名称
    pbottleRPA.log("主流程开始 📍",global_startTime,global_processName)

    await 子流程2('输入参数2');

    console.log('启动子流程  同步子流程'); 
    // 通过require直接执行同步子流程（只包含顶层代码的脚本）
    require('./快速开始演示（3行代码）.js')

    console.log('启动异步子流程 test.js');
    // 调用异步子流程，等待其执行完成并获取返回结果
    // 使用await关键字等待异步操作完成
    let rs = await require('./test.js')('https://rpa.pbottle.com')  // 传递参数给子流程
    console.log('子流程返回结果：',rs);  // 输出子流程返回的结果

    // await process1() //有错误（被注释掉的错误示例）
    pbottleRPA.log("主流程完成 ✅️")  // 在日志中记录主流程完成信息
}

// 执行主流程函数，并捕获可能发生的错误
main().catch((e)=>{  // 主流程错误捕获
    console.log('❌ 未完成，错误',e);  // 输出错误信息
    console.log('准备发送消息给管理员 👉');  // 输出后续处理信息
})

// test.js 文件内容说明（子流程）:
// module.exports = async (url) => {  
//     console.log('子流程开始')  
//     const res = await fetch(url) 
//     pbottleRPA.log('网络请求完成！~',res.ok,res.url)
//     console.log('子流程结束')  
//     return true 
// }


async function 子流程2(params) {  // 定义一个内联的异步函数作为子流程，接收参数，子流程推荐使用中文命名
    console.log('子流程2开始',params);
}

File: python示例/异步子流程模板.py
"""
PBottle RPA asynchronous sub‑process template

Feature description: This script demonstrates how to use sub‑processes (sync/async) in RPA.
Through this example, you can learn how to call synchronous and asynchronous sub‑processes
from the main flow and handle their return values.
"""

# Import the core PBottle RPA library
import pbottleRPA
import subprocess
import sys
import os
import importlib.util

# ========== Global Variables ==========
global_processName = "PBottle RPA — XXXX Process Template"  # Process name
global_startTime = pbottleRPA.getTime()  # Start time


def sync_subprocess(script_path):
    """
    Execute a synchronous sub‑process script (blocking)
    @param script_path: absolute/relative path to a .py file
    """
    print("Launching synchronous sub‑process:", script_path)
    script_path = os.path.abspath(script_path)
    result = subprocess.run(
        [sys.executable, "-u", script_path],
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="ignore",
    )
    if result.returncode == 0:
        print(f"✅ Synchronous sub‑process finished: {script_path}")
        if result.stdout.strip():
            print("  Sub‑process output:", result.stdout.strip())
        return result.stdout
    else:
        print(
            f"❌ Synchronous sub‑process failed (exit code {result.returncode}): {script_path}"
        )
        if result.stderr.strip():
            print("  Error message:", result.stderr.strip())
        return None


def async_subprocess(script_path, *args):
    """
    Run a sub‑process script in a new process (non‑blocking), returns a subprocess object.
    Note: Python's standard library does not directly support async/await style cross‑process calls;
    here we use subprocess + threading to simulate asynchronous behaviour.

    @param script_path: .py file path
    @param *args: arguments passed to the child script via the command line
    @return: subprocess.Popen object, use .wait() to block until completion
    """
    print("Launching asynchronous sub‑process:", script_path)
    script_path = os.path.abspath(script_path)
    proc = subprocess.Popen(
        [sys.executable, "-u", script_path] + list(args),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        encoding="utf-8",
        errors="ignore",
    )
    return proc


def dynamic_import_module(module_path):
    """
    Dynamically import another .py file as a Python module and return it
    (equivalent to JS require() returning module.exports)

    @param module_path: .py file path
    @return: module object whose functions can be called
    """
    module_path = os.path.abspath(module_path)
    module_name = os.path.splitext(os.path.basename(module_path))[0]
    spec = importlib.util.spec_from_file_location(module_name, module_path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


# ========== Sub‑function Definitions ==========
def sub_process2(params):
    """Inline sub‑process function that receives parameters"""
    print("Sub‑process 2 started", params)
    # Write your sub‑process logic here...
    print("Sub‑process 2 ended")
    return True


# ========== Main Flow ==========
def main():
    """Main flow entry point"""
    try:
        # Log main flow start information
        pbottleRPA.log("Main process started 📍", global_startTime, global_processName)

        # ---- Method 1: Call an inline sub‑function directly (synchronous) ----
        print("--- Calling inline sub‑process ---")
        sub_process2("input parameter 2")

        # ---- Method 2: Synchronously invoke an external .py script (blocking) ----
        # Equivalent to JS require('./xxx.js') executing top‑level code synchronously
        print("\n--- Launching synchronous sub‑process ---")
        current_dir = os.path.dirname(os.path.abspath(__file__))
        sync_subprocess(current_dir + "/QuickStart_demo_(3_lines).py")

        # ---- Method 3: Dynamically import an external module and call its functions ----
        # Equivalent to JS let rs = require('./test.js')(url)
        print("\n--- Launching dynamic module import ---")
        try:
            test_module = dynamic_import_module(current_dir + "/download_file_demo.py")
            # Call functions/variables in the module (if it exports any)
            print("✅ Sub‑module loaded successfully:", dir(test_module))
        except Exception as e:
            print(
                f"⚠ Sub‑module load failed (maybe the script is not designed as a module): {e}"
            )

        # ---- Method 4: Asynchronous sub‑process (non‑blocking) ----
        # Equivalent to JS await require('./test.js')() but in non‑blocking form
        print("\n--- Launching asynchronous sub‑process (child process) ---")
        proc = async_subprocess(current_dir + "/QuickStart_demo_(3_lines).py")
        # Do other things here...
        # Wait for the child process to complete
        proc.wait()
        out, err = proc.communicate()
        if out and out.strip():
            print("Asynchronous sub‑process output:", out.strip())

        # Main process finished
        pbottleRPA.log("Main process completed ✅")
        return True

    except Exception as e:
        # Main process error handling
        print("❌ Not completed, error", e)
        print("Preparing to send a message to the administrator 👉")
        # You can add logic here to send a WeChat notification or email:
        # pbottleRPA.wxMessage("RPA process exception", f"Error: {e}", "your_key")
        return False


# ========== Entry Point ==========
if __name__ == "__main__":
    main()
